### PR TITLE
C#: New query VulnerablePackage

### PR DESCRIPTION
**Pull request by calumgrant**
This query finds vulnerable packages imported in project or config files.

Vulnerabilities are described in QL, and the design is extensible to make it straightforward to add new CVEs.

Autobuilder has been changed to also index `.csproj` and `.props` files as XML.

An initial version of this query attempted to use the paths of imported assemblies, but this had the disadvantage that it was hard to report the location of the error and it is often hard to track down the root cause of the import. This new version reports the XML element that caused the import, which is more actionable and easier to test.

### CODE REVIEW

**Comment by jf205** [comment]
@calumgrant I'll have a look at this. Shall I wait for the technical review to be completed first?

**Comment by calumgrant** [comment]
@jf205 That would probably be best.

File: csharp/extractor/Semmle.Autobuild/XmlBuildRule.cs [hunk 1]
```diff
@@ -9,10 +9,16 @@ class XmlBuildRule : IBuildRule
...
                 RunCommand(builder.Odasa).
                 Argument("index --xml --extensions config");
-            return command.Script;
+            var csproj = new CommandBuilder(builder.Actions).
+                RunCommand(builder.Odasa).
+                Argument("index --xml --extensions csproj");
+            var props = new CommandBuilder(builder.Actions).
+                RunCommand(builder.Odasa).
+                Argument("index --xml --extensions props");
+            return config.Script & csproj.Script & props.Script;
```
Comment (hvitved) [inline csharp/extractor/Semmle.Autobuild/XmlBuildRule.cs]: I think you can do it with just one command as `"index --xml --extensions config csproj props"`. Should we perhaps also index `.xml` files while we are at it?

File: csharp/ql/src/Security Features/CWE-937/Vulnerabilities.qll [hunk 2]
```diff
@@ -0,0 +1,290 @@
+/**
+ * A list of NuGet packages with known vulnerabilities.
```
Comment (hvitved) [inline csharp/ql/src/Security Features/CWE-937/Vulnerabilities.qll]: Provides

File: csharp/ql/src/Security Features/CWE-937/Vulnerability.qll [hunk 3]
```diff
@@ -0,0 +1,94 @@
+import csharp
+
+/**
+ * A package reference in an XML file, for example in a
+ * .csproj file, a .props file or a packages.config file.
```
Comment (hvitved) [inline csharp/ql/src/Security Features/CWE-937/Vulnerability.qll]: Comma after file. Add backticks around the extensions?

File: csharp/ql/src/Security Features/CWE-937/Vulnerability.qll [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+ * A vulnerability, where the name of the vulnerability is this string.
+ * One of `matchesRange` or `matchesVersion` must be overridden in order to
+ * specify which packages are vulnerable.
+ */
+abstract class Vulnerability extends string {
+  bindingset[this]
+  Vulnerability() { any() }
+  
+  /**
+   * A package with name `name` is vulnerable from version `affected`
```
Comment (hvitved) [inline csharp/ql/src/Security Features/CWE-937/Vulnerability.qll]: Holds if

File: csharp/ql/src/Security Features/CWE-937/Vulnerability.qll [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+   * until version `fixed`.
+   */
+  predicate matchesRange(string name, Version affected, Version fixed) { none() }
+  
+  /**
+   * A package with name `name` is vulnerable in version `affected`, and
```
Comment (hvitved) [inline csharp/ql/src/Security Features/CWE-937/Vulnerability.qll]: Holds if

File: csharp/ql/src/Security Features/CWE-937/Vulnerability.qll [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+   * is fixed by version `fixed`.
+   */
+  predicate matchesVersion(string name, Version affecter, Version fixed) { none() }
```
Comment (hvitved) [inline csharp/ql/src/Security Features/CWE-937/Vulnerability.qll]: `affecter` -> `affected`

File: csharp/ql/src/Security Features/CWE-937/Vulnerabilities.qll [hunk 7]
```diff
@@ -0,0 +1,290 @@
...
+      name = "Microsoft.AspNetCore.Server.Kestrel.Core"
+      or
+      name = "Microsoft.AspNetCore.Server.Kestrel.Transport.Abstractions"
+      or
+      name = "Microsoft.AspNetCore.Server.Kestrel.Transport.Libuv"
+      or
+      name = "Microsoft.AspNetCore.All"
+    ) and
+    affected = "2.0.0" and
+    fixed = "2.0.3"
```
Comment (hvitved) [inline csharp/ql/src/Security Features/CWE-937/Vulnerabilities.qll]: `fixed` should be `"2.0.8` for `"Microsoft.AspNetCore.All"`.

File: csharp/ql/src/Security Features/CWE-937/VulnerablePackage.qhelp [hunk 8]
```diff
@@ -0,0 +1,43 @@
+<!DOCTYPE qhelp PUBLIC
+  "-//Semmle//qhelp//EN"
+  "qhelp.dtd">
+<qhelp>
+
+<overview>
+<p>
+Using a package with a known vulnerability is a security risk that could leave the
+software vulnerable to attack.
```
Comment (hvitved) [inline csharp/ql/src/Security Features/CWE-937/VulnerablePackage.qhelp]: an attack or attacks?

File: csharp/ql/src/Security Features/CWE-937/VulnerablePackage.ql [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+ * @tags security
+ *       external/cwe/cwe-937
+ */
+
+import csharp
+import Vulnerabilities
+
+from Vulnerability vuln, VulnerablePackage package
+where vuln = package.getVulnerability()
+select package, "Package " + package + " has vulnerability $@, and should be upgraded to version " + package.getFixedVersion() + ".",
```
Comment (hvitved) [inline csharp/ql/src/Security Features/CWE-937/VulnerablePackage.ql]: add `'` around package

File: csharp/ql/src/Security Features/CWE-937/VulnerablePackageGOOD.csproj [hunk 10]
```diff
@@ -0,0 +1,15 @@
+<Project Sdk="Microsoft.NET.Sdk">
```
Comment (hvitved) [inline csharp/ql/src/Security Features/CWE-937/VulnerablePackageGOOD.csproj]: Copy this file (and `VulnerablePackageBAD.csproj`) to the test.

File: csharp/ql/src/Security Features/CWE-937/VulnerablePackage.ql (hunk 9 above)
Comment (hvitved) [reply csharp/ql/src/Security Features/CWE-937/VulnerablePackage.ql]: Have you checked that the links to `vuln.getUrl()` render correctly in both QL4E and on LGTM.com?

File: csharp/ql/src/Security Features/CWE-937/VulnerablePackageGOOD.csproj (hunk 10 above)
Comment (calumgrant) [reply csharp/ql/src/Security Features/CWE-937/VulnerablePackageGOOD.csproj]: Currently, qltest doesn't extract `.csproj` files as XML so it wouldn't do anything. I'm not really sure it's worth making that change.

Comment (hvitved) [reply csharp/ql/src/Security Features/CWE-937/VulnerablePackageGOOD.csproj]: OK.

File: csharp/ql/src/Security Features/CWE-937/VulnerablePackage.qhelp (hunk 8 above)
Comment (jf205) [reply csharp/ql/src/Security Features/CWE-937/VulnerablePackage.qhelp]: I think you could use any of the above suggestions–my preference would be to leave as it is though.

File: csharp/ql/src/Security Features/CWE-937/VulnerablePackageGOOD.csproj (hunk 10 above)
Comment (hvitved) [reply csharp/ql/src/Security Features/CWE-937/VulnerablePackageGOOD.csproj]: OK.

Comment (hvitved) [reply csharp/ql/src/Security Features/CWE-937/VulnerablePackageGOOD.csproj]: OK.

File: csharp/ql/src/Security Features/CWE-937/VulnerablePackage.qhelp (hunk 8 above)
Comment (jf205) [reply csharp/ql/src/Security Features/CWE-937/VulnerablePackage.qhelp]: I think you could use any of the above suggestions–my preference would be to leave as it is though.

File: csharp/ql/src/Security Features/CWE-937/VulnerablePackage.qhelp [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+This query reads the packages imported by the project build files and
+<code>.config</code> files, and checks them against a list of packages with known
+vulnerabilities.
+</p>
+</overview>
+
+<recommendation>
+<p>
+Upgrade the package to the recommended version, for example using the NuGet package manager,
+or by editing the project files directly.
```
Comment (jf205) [inline csharp/ql/src/Security Features/CWE-937/VulnerablePackage.qhelp]: Suggest very minor rewording here:
Upgrade the package to the recommended version, using, for example, the NuGet package manager, or by editing the project files directly.

Comment (jf205) [inline csharp/ql/src/Security Features/CWE-937/VulnerablePackage.qhelp]: Suggest very minor rewording here:
Upgrade the package to the recommended version, using, for example, the NuGet package manager, or by editing the project files directly.

File: csharp/ql/src/Security Features/CWE-937/VulnerablePackage.qhelp (hunk 8 above)
Comment (jf205) [reply csharp/ql/src/Security Features/CWE-937/VulnerablePackage.qhelp]: Any of the above would work I think–my preference would be to leave as is.

File: csharp/ql/src/Security Features/CWE-937/VulnerablePackage.qhelp (hunk 11 above)
Comment (jf205) [inline csharp/ql/src/Security Features/CWE-937/VulnerablePackage.qhelp]: Suggest a very minor re-wording here:
Upgrade the package to the recommended version using, for example, the NuGet package manager, or by editing the project files directly.

Comment (jf205) [inline csharp/ql/src/Security Features/CWE-937/VulnerablePackage.qhelp]: Suggest a very minor re-wording here:
Upgrade the package to the recommended version using, for example, the NuGet package manager, or by editing the project files directly.

File: csharp/ql/src/Security Features/CWE-937/VulnerablePackage.ql (hunk 9 above)
Comment (hvitved) [reply csharp/ql/src/Security Features/CWE-937/VulnerablePackage.ql]: @calumgrant : Did you check this?

Comment (hvitved) [reply csharp/ql/src/Security Features/CWE-937/VulnerablePackage.ql]: If the above does not work, I think something like this should:
```
"Package " + package + " has vulnerability [[\"" + vuln + "\"|\"" +vuln.getUrl()+ "\"]], and should be upgraded to version " + package.getFixedVersion() + "."
```

Comment (hvitved) [reply csharp/ql/src/Security Features/CWE-937/VulnerablePackage.ql]: If the above does not work, I think something like this should:
```
"Package " + package + " has vulnerability [[\"" + vuln + "\"|\"" +vuln.getUrl()+ "\"]], and should be upgraded to version " + package.getFixedVersion() + "."
```

Comment (hvitved) [reply csharp/ql/src/Security Features/CWE-937/VulnerablePackage.ql]: If the above does not work, I think something like this should:
```
"Package " + package + " has vulnerability [[\"" + vuln + "\"|\"" +vuln.getUrl()+ "\"]], and should be upgraded to version " + package.getFixedVersion() + "."
```

**Pull request merged**